android/filament-android/src/main/java/com/google/android/filament/View.java

```diff
@@ -1548,6 +1548,63 @@ public DepthOfFieldOptions getDepthOfFieldOptions() {
         return mDepthOfFieldOptions;
     }
 
+    /**
+     * A class containing the result of a picking query
+     */
+    static class PickingQueryResult {
+        /** The entity of the renderable at the picking query location */
+        @Entity int renderable;
+        /** The value of the depth buffer at the picking query location */
+        float depth;
+        /** The fragment coordinate in GL convention at the the picking query location */
+        @NonNull float[] fragCoords = new float[3];
+    };
+
+    public interface OnPickCallback {
+        void onPick(@NonNull PickingQueryResult result);
+    }
+
+    /**
+     * Creates a picking query. Multiple queries can be created (e.g.: multi-touch).
+     * Picking queries are all executed when {@link Renderer#render} is called on this View.
+     * The provided callback is guaranteed to be called at some point in the future.
+     *
+     * Typically it takes a couple frames to receive the result of a picking query.
+     *
+     * @param x        Horizontal coordinate to query in the viewport with origin on the left.
+     * @param y        Vertical coordinate to query on the viewport with origin at the bottom.
+     * @param handler  An {@link java.util.concurrent.Executor Executor}.
+     *                 On Android this can also be a {@link android.os.Handler Handler}.
+     * @param callback User callback executed by <code>handler</code> when the picking query
+     *                 result is available.
+     */
+    public void pick(int x, int y,
+            @Nullable Object handler, @Nullable OnPickCallback callback) {
+        InternalOnPickCallback internalCallback = new InternalOnPickCallback(callback);
+        nPick(getNativeObject(), x, y, handler, internalCallback);
+    }
+
+    private static class InternalOnPickCallback implements Runnable {
+        public InternalOnPickCallback(OnPickCallback mUserCallback) {
+            this.mUserCallback = mUserCallback;
+        }
+        @Override
+        public void run() {
+            mPickingQueryResult.renderable = mRenderable;
+            mPickingQueryResult.depth = mDepth;
+            mPickingQueryResult.fragCoords[0] = mFragCoordsX;
+            mPickingQueryResult.fragCoords[1] = mFragCoordsY;
+            mPickingQueryResult.fragCoords[2] = mFragCoordsZ;
+            mUserCallback.onPick(mPickingQueryResult);
+        }
+        private final OnPickCallback mUserCallback;
+        private final PickingQueryResult mPickingQueryResult = new PickingQueryResult();
+        @Entity public int mRenderable;
+        public float mDepth;
```

Comment:
You don't need those fields, the JNI could directly set the values in `mPickingQueryResult`.

Reply:
Yes, it seemed harder to do, because I would have to retrieve the object's field from jni, or pass it to the jni, and then I would have to capture it in the lambda. And technically, I was wondering if I would have to take a global ref on it (it just rely on the fact that there is one on the object that owns it).
Just seemed more complicated :-)

Reply:
Yeah but that's no different from what you do now. Except on a different class.

Reply:
I was going to go down this route, and then realized that I would need to manage the array of 3 floats from JNI, so decided to keep it simple for now.

Reply:
Using an array from JNI is easier than looking up/setting 3 fields :) Anyway it doesn't matter much, it would just make the code simpler.